Add standalon rest assured test

kovaku opened the pull request.

fszaboBEX committed:
Initial standalone tests

fszaboBEX committed:
Introduce jsoncompare, parametrized user test

fszaboBEX committed:
Springboot IT test prototype

fszaboBEX committed:
Added negative cases for get user by id

fszaboBEX committed:
Added user by name retrieval scenarios

fszaboBEX committed:
Added new record creation scenario

kovaku commented on src/test/java/com/github/kovaku/user/GetAllUserTest.java:
```diff
@@ -0,0 +1,51 @@
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class GetAllUserTest {
```
By default we use the public access modifier on classes, since your class is currently package private.
More on this: https://www.javatpoint.com/access-modifiers

lammaer replied:
Fixed

kovaku commented on standalone-e2e-test/pom.xml:
```diff
@@ -0,0 +1,68 @@
+  </properties>
+  <name>standalone-e2e-test</name>
+  <url>http://maven.apache.org</url>
+  <dependencies>
```
Please add some blank lines between the main sections of the pom, so that will be more readable.

lammaer replied:
Fixed

kovaku commented on standalone-e2e-test/pom.xml:
```diff
@@ -0,0 +1,68 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>org.example</groupId>
```
Could be com.github.lammaer. Naming conventions: https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html

lammaer replied:
Fixed

kovaku commented on standalone-e2e-test/pom.xml:
```diff
@@ -0,0 +1,68 @@
+              <goal>write-project-properties</goal>
+            </goals>
+            <configuration>
+              <outputFile>${project.build.outputDirectory}/properties-from-pom.properties</outputFile>
```
I cannot find this file, what do you want to achieve with this config?

lammaer replied:
It was an experiment to implement environment specific property files following this guide
https://www.baeldung.com/java-accessing-maven-properties

but then I followed a different route, so not used it but remained in the pom. Removed.

kovaku commented on standalone-e2e-test/src/test/java/AbstractRestAssuredBase.java:
```diff
@@ -0,0 +1,37 @@
+    @BeforeAll
+    public static void init() {
+        String targetenv = System.getProperty("targetenv");
+        System.out.println("environment is " + targetenv);
```
Please use logback as a logging library, follow this guide: https://www.baeldung.com/logback

As I remember RestAssured has build in logger, so don't worry about that one, just log some useful messages with logback.

lammaer replied:
Done

kovaku commented on src/test/java/com/github/kovaku/user/GetAllUserTest.java:
```diff
@@ -0,0 +1,51 @@
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class GetAllUserTest {
+    @LocalServerPort
+    int randomServerPort;
```
Fields should be private.

lammaer replied:
Fixed

kovaku commented on src/test/java/com/github/kovaku/user/GetAllUserTest.java:
```diff
@@ -0,0 +1,51 @@
+    int randomServerPort;
+    URI uri;
+
+    @Value("classpath:alldefaultusers.json")
```
Nice

kovaku commented on src/test/java/com/github/kovaku/user/GetAllUserTest.java:
```diff
@@ -0,0 +1,51 @@
+    URI uri;
+
+    @Value("classpath:alldefaultusers.json")
+    Resource alldefaultusersresource;
```
Don't forget the camelCase format.

lammaer replied:
Fixed

kovaku commented on src/test/java/com/github/kovaku/user/GetAllUserTest.java:
```diff
@@ -0,0 +1,51 @@
+
+    @BeforeAll
+    void init () throws URISyntaxException {
+    uri = new URI("http://localhost:" + randomServerPort + "/api/v1/users");
```
Missing indentation.

lammaer replied:
Fixed

kovaku commented on src/test/java/com/github/kovaku/user/GetAllUserTest.java:
```diff
@@ -0,0 +1,51 @@
+
+        Assertions.assertEquals(200, result.getStatusCodeValue());
+
+        String usersExpected = new String(Files.readAllBytes(alldefaultusersresource.getFile().toPath()));
```
If I were you, I would simply use: https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/IOUtils.html#resourceToString-java.lang.String-java.nio.charset.Charset-

lammaer replied:
This is great, much simpler, thx!

kovaku commented on src/test/java/com/github/kovaku/user/GetAllUserTest.java:
```diff
@@ -0,0 +1,51 @@
+import java.nio.file.Files;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```
Nice, an IT test!

kovaku commented on src/test/java/com/github/kovaku/user/GetAllUserTest.java:
```diff
@@ -0,0 +1,51 @@
+
+        String usersExpected = new String(Files.readAllBytes(alldefaultusersresource.getFile().toPath()));
+        usersExpected = usersExpected.replace("8080", String.valueOf(randomServerPort));  //because the response URL contains the port what is random
+        JSONAssert.assertEquals(usersExpected,result.getBody(), JSONCompareMode.STRICT );
```
You should leave a whitespace after all the commas.

kovaku replied:
The autoformatter is available using Ctrl+Alt+L in Intellij Idea.

lammaer replied:
Readability is for the weaklings! ;) Fixed

kovaku commented on src/test/java/com/github/kovaku/user/GetAllUserTest.java:
```diff
@@ -0,0 +1,51 @@
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.web.server.LocalServerPort;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
```
I think this is not used, use Ctrl+Alt+O to optimize imports in Intellij Idea.

lammaer replied:
Your were right, enabled "Optimize imports on the fly" in intellij settings.

kovaku commented on standalone-e2e-test/src/test/java/GetAllUsersTest.java:
```diff
@@ -0,0 +1,75 @@
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class GetAllUsersTest extends AbstractRestAssuredBase {
```
You can simply convert these tests to use spring using the @ExtendsWith annotation.
Sample: https://www.concretepage.com/spring-5/extendwith-springextension-spring-test
In the class used in the @ContextConfiguration annotation you can use @ComponentScan and also @PropertySource to load properties from files.

lammaer replied:
Done

kovaku commented on standalone-e2e-test/src/test/java/GetAllUsersTest.java:
```diff
@@ -0,0 +1,75 @@
+                .extract()
+                .asString();
+
+        JsonPath js = new JsonPath(responseString);
```
You could also try to deserialize the response as an java object.
https://www.toolsqa.com/rest-assured/deserialize-json-response/

lammaer replied:
Added test class testDefaultGetAllUsersWithDeserializedJavaObject to demonstrate this method

kovaku commented on standalone-e2e-test/src/test/java/GetUserByIdTest.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @ParameterizedTest
+    @MethodSource("defaultUserNames")
```
Nice

kovaku commented on standalone-e2e-test/README.MD:
```diff
@@ -0,0 +1,10 @@
+mvn clean install -f standalone-e2e-test/pom.xml
```
Instead of putting an additional maven modul inside the existing one, maybe you can try to refactor the whole project to a real multi module maven project: 
* Theory: https://www.baeldung.com/maven-multi-module
* In practice: https://mkyong.com/maven/maven-how-to-create-a-multi-module-project/

lammaer replied:
Changed. Thanks for the PAIN :D

fszaboBEX committed:
Refactor round #1 on PR comments
- readability fixes, removing unused imports
- use org.apache.commons.io.IOUtils.resourceToString

fszaboBEX committed:
Added logback to E2E tests

fszaboBEX committed:
Added logback to E2E tests

fszaboBEX committed:
Restructuring to maven multi module project (note to self: the service itself had to be moved to simple submodule as well)

fszaboBEX committed:
Non working attempt to introduce retrieving config from property files

fszaboBEX committed:
Introducing retrieval parameters from env specific property files using spring

fszaboBEX committed:
Simplifying execution command line options with maven profiles

fszaboBEX committed:
Adding an example how to deserialise JSON response to a POJO

fszaboBEX committed:
Cleanup dependencies in E2E module

fszaboBEX committed:
Set logging to INFO since springboot starters flood the log in DEBUG mode

kovaku commented on pom.xml:
```diff
@@ -1,75 +1,148 @@
+                <dependency>
+                    <groupId>io.rest-assured</groupId>
+                    <artifactId>rest-assured</artifactId>
+                    <version>5.2.0</version>
```
You could extract the common versions of multiple dependencies that are tied together.

kovaku commented on standalone-e2e-test/pom.xml:
```diff
@@ -0,0 +1,94 @@
+  <version>1.0-SNAPSHOT</version>
+  
+  <parent>
+	 <groupId>com.github.kovaku</groupId>
```
The indentation is not consistent in this file, maybe you could use checkstyle in your project.

kovaku commented on standalone-e2e-test/pom.xml:
```diff
@@ -0,0 +1,94 @@
+    </dependency>
+  </dependencies>
+
+  <profiles>
```
Nice.

kovaku commented on standalone-e2e-test/src/test/java/AbstractRestAssuredBase.java:
```diff
@@ -0,0 +1,32 @@
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = RestAssuredTestConfiguration.class)
+public class AbstractRestAssuredBase {
+    static final Logger log = LoggerFactory.getLogger(AbstractRestAssuredBase.class);
```
Since this is a constant, the convention is to write in UPPER_CASE format with underscore as separator.

kovaku commented on standalone-e2e-test/src/test/java/AbstractRestAssuredBase.java:
```diff
@@ -0,0 +1,32 @@
+        return RestAssured.given().baseUri(baseUri);
+    }
+
+    public static String getJsonStringFromFile(String filename) throws IOException {
```
Just for fun, an for the sake of practicing spring, you should create a utility class as a bean and inject it into this class.
You can use component scan to discover your new bean or you can define it as a bean in the configuration class, I leave it to you to test these out and pick one solution.

FYI: @Value annotation can be used to load resources as well, you can play with that as well.

kovaku commented on standalone-e2e-test/src/test/java/AbstractRestAssuredBase.java:
```diff
@@ -0,0 +1,32 @@
+import configuration.RestAssuredTestConfiguration;
```
I just realized that you haven't used some packages for your classes. Please use something like com.github.lammaer as like your group id of the project.

kovaku commented on standalone-e2e-test/src/test/java/GetAllUsersTest.java:
```diff
@@ -0,0 +1,93 @@
+                .get();
+
+        ResponseBody actualBody = actualResponse.getBody();
+        List<UserResponse> userResponse = Arrays.asList(actualBody.as(UserResponse[].class));
```
Is nice.

kovaku commented on standalone-e2e-test/src/test/java/GetAllUsersTest.java:
```diff
@@ -0,0 +1,93 @@
+        Assertions.assertEquals(1, userResponse.get(0).id);
+        Assertions.assertEquals("John Doe", userResponse.get(0).name);
+    }
+}
```
Please add one additional test where you try JsonAssert as well.
https://www.baeldung.com/jsonassert

kovaku commented on standalone-e2e-test/src/test/java/GetUserByIdTest.java:
```diff
@@ -0,0 +1,63 @@
+                .get()
+                .prettyPeek()
+                .then()
+                .statusCode(404);
```
You can use some predefined constants provided by Spring:
HttpStatus.NOT_FOUND.value();

kovaku commented on standalone-e2e-test/src/test/java/configuration/RestAssuredTestConfiguration.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.context.annotation.PropertySource;
+
+@Configuration
+@ComponentScan("config")
```
I don't think that you have any root package called config, there spring won't find any bean for you. You can remove it.

kovaku commented on standalone-e2e-test/src/test/java/response/Links.java:
```diff
@@ -0,0 +1,6 @@
+package response;
+
```
POJOs are usually encapsulated, meaning that their field are private and can be accessed only by setters and getters.
